Ra off binary formatter (#19445)
dotnet/docs: docs/standard/serialization/binaryformatter-security-guide.md

---
title: "BinaryFormatter security guide"
description: This article describes the security risks inherent in the BinaryFormatter type and recommendations for different serializers to use.
ms.date: "07/11/2020"
ms.author: levib
author: GrabYourPitchforks
---
# BinaryFormatter security guide

> This document applies to the following .NET implementations:

Reviewer: ```suggestion
This article applies to the following .NET implementations:
```

* .NET Framework all versions
* .NET Core 2.1 - 3.1
* .NET 5.0 and later

## Background

> [!WARNING]
> The <xref:System.Runtime.Serialization.Formatters.Binary.BinaryFormatter> type is dangerous and is ***not*** recommended for data processing. Applications should stop using `BinaryFormatter` as soon as possible, even if they believe the data they're processing to be trustworthy. `BinaryFormatter` is insecure can't be made secure.

> This document also applies to the following types:

Reviewer: ```suggestion
This article also applies to the following types:
```

In .NET, the biggest risk target is apps that use the <xref:System.Runtime.Serialization.Formatters.Binary.BinaryFormatter> type to deserialize data. `BinaryFormatter` is widely used throughout the .NET ecosystem because of its power and its ease of use. However, this same power gives attackers the ability to influence control flow within the target app. Successful attacks can result in the attacker being able to run code within the context of the target process.

As a simpler analogy, assume that calling `BinaryFormatter.Deserialize` over a payload is the equivalent of interpreting that payload as a standalone executable and launching it.

## BinaryFormatter security vulnerabilities

> [!WARNING]

> > The `BinaryFormatter.Deserialize` method is __never__ safe when used with untrusted input. We strongly recommend that consumers instead consider using one of the alternatives outlined later in this document.

Reviewer: ```suggestion
> The `BinaryFormatter.Deserialize` method is __never__ safe when used with untrusted input. We strongly recommend that consumers instead consider using one of the alternatives outlined later in this article.
```

> The implementation of `BinaryFormatter` was originally created before deserialization vulnerabilities were a well-understood threat category. As a result, the code does not follow modern best practices. The `Deserialize` method can be used as a vector for attackers to perform DoS attacks against consuming apps. These attacks might render the app unresponsive or result in unexpected process termination. This category of attack cannot be mitigated with a `SerializationBinder` or any other `BinaryFormatter` configuration switch. .NET considers this behavior ***by design*** and won't issue a code update to modify the behavior.

Reviewer: ```suggestion
`BinaryFormatter` was implemented before deserialization vulnerabilities were a well-understood threat category. As a result, the code does not follow modern best practices. The `Deserialize` method can be used as a vector for attackers to perform DoS attacks against consuming apps. These attacks might render the app unresponsive or result in unexpected process termination. This category of attack cannot be mitigated with a `SerializationBinder` or any other `BinaryFormatter` configuration switch. .NET considers this behavior to be ***by design*** and won't issue a code update to modify the behavior.
```

`BinaryFormatter.Deserialize` may be vulnerable to other attack categories, such as information disclosure or remote code execution. Utilizing features such as a custom <xref:System.Runtime.Serialization.SerializationBinder> may be insufficient to properly mitigate these risks. The possibility exists that a novel vulnerability will be discovered for which .NET cannot practically publish a security update. Consumers should assess their individual scenarios and consider their potential exposure to these risks.

> We recommend that `BinaryFormatter` consumers perform individual risk assessments on their apps. It is the consumer's sole responsibility to determine whether to utilize `BinaryFormatter`. Consumers should risk-assess security, technical, reputation, legal, and regulatory requirements of using `BinaryFormatter`.

Reviewer: ```suggestion
We recommend that `BinaryFormatter` consumers perform individual risk assessments on their apps. It is the consumer's sole responsibility to determine whether to utilize `BinaryFormatter`. Consumers should risk assess the security, technical, reputation, legal, and regulatory requirements of using `BinaryFormatter`.
```

## Preferred alternatives

.NET offers several in-box serializers that can handle untrusted data safely:

* <xref:System.Xml.Serialization.XmlSerializer> and <xref:System.Runtime.Serialization.DataContractSerializer> to serialize object graphs into and from XML. Do not confuse `DataContractSerializer` with  <xref:System.Runtime.Serialization.NetDataContractSerializer>.

> <!-- Levi comment left in your original PR applied here-->

Reviewer: ```suggestion
```
